theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,292 @@
+            ctc_context_destroy( context, 1 );
+
+            PyErr_Format( PyExc_MemoryError,
+                "Could not allocate storage for CTC costs" );
+            return 1;
+        }
+    }
+
+    costs = (npy_float32 *) PyArray_DATA( *out_costs );
+
+    npy_float32 * gradients = NULL;
+
+    if ( NULL != out_gradients )  // If gradient computation is not disabled
+    {
+        if ( NULL == (*out_gradients) ||  // Symbolic variable has no real backing
+            PyArray_NDIM( *out_gradients ) != 3 ||
+            PyArray_DIMS( *out_gradients )[0] != PyArray_DIMS( in_activations )[0] ||
+            PyArray_DIMS( *out_gradients )[1] != PyArray_DIMS( in_activations )[1] ||
+            PyArray_DIMS( *out_gradients )[2] != PyArray_DIMS( in_activations )[2] )
+        {
+            // Existing matrix is the wrong size. Make a new one.
+            // Decrement ref counter to existing array
+            Py_XDECREF( *out_gradients );
+            // Allocate new array
+            *out_gradients = (PyArrayObject *) PyArray_ZEROS(3, PyArray_DIMS( in_activations ),
+                NPY_FLOAT32, 0);
+            // Add a copy of the newly allocated gradients, so that we can
+            // decrease its reference (deallocate it) in case of errors inside
+            // the wrapper.
+            context->gradients = *out_gradients;
+
+            if ( NULL == (*out_gradients) )
+            {
+                // Destroy previous CTC context before returning exception
+                ctc_context_destroy( context, 1 );
+
+                PyErr_Format( PyExc_MemoryError,
+                    "Could not allocate storage for CTC gradients!" );
+                return 1;
+            }
+        }
+        gradients = (npy_float32 *) PyArray_DATA( *out_gradients );
+    }
+
+    size_t cpu_workspace_size;
+    int ctc_error;
+
+    ctc_error = ctc_check_result( get_workspace_size( context->label_lengths,
+        context->input_lengths, alphabet_size, minibatch_size, context->options,
+        &cpu_workspace_size ),
+        "Failed to obtain CTC workspace size!" );
+
+    if ( ctc_error )  // Exception is set by ctc_check_result, return error here
+        return 1;
+
+    context->workspace = malloc( cpu_workspace_size );
+
+    if ( NULL == context->workspace )
+    {
+        // Destroy previous CTC context before returning exception
+        ctc_context_destroy( context, 1 );
+
+        PyErr_Format( PyExc_MemoryError,
+            "Failed to allocate memory for CTC workspace!" );
+        return 1;
+    }
+
+    ctc_error = ctc_check_result( compute_ctc_loss( activations, gradients,
+        context->flat_labels, context->label_lengths, context->input_lengths,
+        alphabet_size, minibatch_size, costs, context->workspace,
+        context->options ), "Failed to compute CTC loss function!" );
+
+    if ( ctc_error )  // Exception is set by ctc_check_result, return error here
+        return 1;
```

Comment:
We should call `ctc_context_destroy()` before returning here, as ctc context may contain allocated memory.